### PR TITLE
Add doc for CA2020

### PR DESCRIPTION
**Pull request by buyaa-n**
## Summary

Add doc for new rule CA2020

Related to https://github.com/dotnet/runtime/issues/74022

The analyzer PR: https://github.com/dotnet/roslyn-analyzers/pull/6153

### CODE REVIEW

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+| explicit operator IntPtr(void\*) | checked | throws when overflows | do not throw when overflow | `checked((IntPtr)voidPtrVariable);` |
+| explicit operator int(IntPtr) | unchecked | will not throw when overflows | could throw in 64-bit contexts | `(int)intPtrVariable;` |
+| operator +(UIntPtr, int) | checked | throws when overflows | do not throw when overflows | `checked(uintPtrVariable + 2);` |
+| operator -(UIntPtr, int) | checked | throws when overflows | do not throw when overflows | `checked(uintPtrVariable - 2);` |
+| explicit operator UIntPtr(ulong) | unchecked | will not throw when overflows | could throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
+| explicit operator uint(UIntPtr) | unchecked | will not throw when overflows | could throw in 64-bit contexts | `(uint)uintPtrVariable` |
+
+## How to fix violations
+
+Investigate the code to determine if the expression flagged could cause a behavioral change, and choose appropriate way to fix the diagnostic
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:47]:
```suggestion
Examine your code to determine if the flagged expression could cause a behavioral change, and choose an appropriate way to fix the diagnostic from the following options:
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          intPtrVariable = checked((IntPtr)longVariable); // wrap with checked
+
+          int a = checked((int)intPtrVariable); //
+      }
+  }
+  ```
+
+## When to suppress warnings
+
+If the expression would not cause a behavioral change it is safe to suppress a warning from this rule.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:154]:
```suggestion
If the expression would not cause a behavioral change, it's safe to suppress a warning from this rule.
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+          }
+
+          intPtrVariable = (nint)longVariable; 
+
+          int a = (int)intPtrVariable;
+      }
+  }
+  ```
+
+- If the expression could cause a behavioral change wrap it with corresponding `checked`/`unchecked` statement to preserve the old .NET 6.0 behavior.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:122]:
```suggestion
- If the expression could cause a behavioral change, wrap it with a `checked` or `unchecked` statement to preserve the previous behavior.
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        intPtrVariable = (IntPtr)longVariable; // Starting with .NET 7 the explicit conversion '(IntPtr)Int64' will not throw when overflowing in an unchecked context. Wrap the expression with a 'checked' statement to restore the .NET 6 behavior.
+
+        int a = (int)intPtrVariable; // Starting with .NET 7 the explicit conversion '(Int32)IntPtr' will not throw when overflowing in an unchecked context. Wrap the expression with a 'checked' statement to restore the .NET 6 behavior.
+    }
+}
+```
+
+**Fix**:
+
+- If the expression would not cause a behavioral change:
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:91]:
```suggestion
- If the expression would not cause a behavioral change and the the `IntPtr` or `UIntPtr` type is used as a native `int` or `uint`, change the type to `nint` or `nuint`.
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  - If the IntPtr/UIntPtr type used as native int/uint purpose change the type to nint/nuint.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:92]:
```suggestion
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+Investigate the code to determine if the expression flagged could cause a behavioral change, and choose appropriate way to fix the diagnostic
+
+Fix options:
+
+- If the expression would not cause a behavioral change:
+  - If the `IntPtr/UIntPtr` type used as native int/uint purpose change the type to `nint/nuint`.
+  - If the `IntPtr/UIntPtr` type used as native pointer change the type to corresponding native pointer type
+  - If you cannot change the type of the variable last resort will be suppress the warning
+- If the expression could cause a behavioral change wrap it with corresponding `checked`/`unchecked` statement to preserve the .NET 6.0 behavior.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:55]:
```suggestion
- If the expression could cause a behavioral change, wrap it with a `checked` or `unchecked` statement to preserve the previous behavior.
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+## How to fix violations
+
+Investigate the code to determine if the expression flagged could cause a behavioral change, and choose appropriate way to fix the diagnostic
+
+Fix options:
+
+- If the expression would not cause a behavioral change:
+  - If the `IntPtr/UIntPtr` type used as native int/uint purpose change the type to `nint/nuint`.
+  - If the `IntPtr/UIntPtr` type used as native pointer change the type to corresponding native pointer type
+  - If you cannot change the type of the variable last resort will be suppress the warning
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:54]:
```suggestion
  - If you can't change the type of the variable, suppress the warning.
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+## How to fix violations
+
+Investigate the code to determine if the expression flagged could cause a behavioral change, and choose appropriate way to fix the diagnostic
+
+Fix options:
+
+- If the expression would not cause a behavioral change:
+  - If the `IntPtr/UIntPtr` type used as native int/uint purpose change the type to `nint/nuint`.
+  - If the `IntPtr/UIntPtr` type used as native pointer change the type to corresponding native pointer type
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:53]:
```suggestion
  - If the `IntPtr` or `UIntPtr` type is used as a native pointer, change the type to the corresponding native pointer type.
```

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+| explicit operator uint(UIntPtr) | unchecked | will not throw when overflows | could throw in 64-bit contexts | `(uint)uintPtrVariable` |
+
+## How to fix violations
+
+Investigate the code to determine if the expression flagged could cause a behavioral change, and choose appropriate way to fix the diagnostic
+
+Fix options:
+
+- If the expression would not cause a behavioral change:
+  - If the `IntPtr/UIntPtr` type used as native int/uint purpose change the type to `nint/nuint`.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:52]:
```suggestion
  - If the `IntPtr` or `UIntPtr` type is used as a native `int` or `uint`, change the type to `nint` or `nuint`.
```

**Review by gewarren — APPROVED** [review]
Thanks @buyaa-n!

File: docs/fundamentals/code-analysis/quality-rules/ca2020.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+| explicit operator IntPtr(long) | Unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(IntPtr)longVariable;` |
+| explicit operator void\*(IntPtr) | Checked | throws when overflows | Doesn't throw when overflows | `checked((void*)intPtrVariable);` |
+| explicit operator IntPtr(void\*) | Checked | throws when overflows | Doesn't throw when overflows | `checked((IntPtr)voidPtrVariable);` |
+| explicit operator int(IntPtr) | Unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(int)intPtrVariable;` |
+| operator +(UIntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable + 2);` |
+| operator -(UIntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable - 2);` |
+| explicit operator UIntPtr(ulong) | Unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
+| explicit operator uint(UIntPtr) | Unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(uint)uintPtrVariable` |
```
Comment (buyaa-n) [inline docs/fundamentals/code-analysis/quality-rules/ca2020.md:43]: I think Checked/Unchecked better lower cased same as the checked/unchecked keywords

```suggestion
| operator +(IntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable + 2);` |
| operator -(IntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable - 2);` |
| explicit operator IntPtr(long) | unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(IntPtr)longVariable;` |
| explicit operator void\*(IntPtr) | checked | throws when overflows | Doesn't throw when overflows | `checked((void*)intPtrVariable);` |
| explicit operator IntPtr(void\*) | checked | throws when overflows | Doesn't throw when overflows | `checked((IntPtr)voidPtrVariable);` |
| explicit operator int(IntPtr) | unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(int)intPtrVariable;` |
| operator +(UIntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable + 2);` |
| operator -(UIntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable - 2);` |
| explicit operator UIntPtr(ulong) | unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
| explicit operator uint(UIntPtr) | unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(uint)uintPtrVariable` |
```

**Pull request merged**